codegen.ts

```diff
@@ -1,22 +1,78 @@
-import { run } from "./codegen/mod.ts";
+import { codegen } from "./codegen/mod.ts";
+import { Config } from "./config/mod.ts";
 import { parse } from "./deps/std/flags.ts";
+import * as path from "./deps/std/path.ts";
+import { unimplemented } from "./deps/std/testing/asserts.ts";
+import * as M from "./frame_metadata/mod.ts";
+import { proxyClient } from "./rpc/providers/proxy.ts";
+import * as U from "./util/mod.ts";
 
 const args = parse(Deno.args, {
-  string: ["metadata", "output"],
+  string: ["src", "out", "import"],
   boolean: ["help"],
+  default: {
+    import: "deno",
+  },
   alias: {
-    metadata: ["m"],
-    output: ["out", "o"],
+    src: ["s"],
+    out: ["o"],
     help: ["h", "?"],
   },
 });
 
-if (args.help) {
-  console.log("Usage: codegen --metadata <file> --output <dir>");
-  Deno.exit(0);
+if (args.help) help();
+if (!args.src) fail();
+if (!args.out) fail();
+
+await codegen({
+  importSpecifier: args.import === "npm"
+    ? "capi"
+    : args.import === "deno"
+    ? "https://deno.land/x/capi/mod.ts"
+    : args.import,
+  metadata: await getMetadata(args.src),
+  outDir: args.out,
+}).write();
+
+async function getMetadata(src: string): Promise<M.Metadata> {
+  if (src.startsWith("ws")) {
+    const client = U.throwIfError(await proxyClient(new Config(() => src, undefined!)));
+    const metadata = U.throwIfError(await client.call("state_getMetadata", []));
+    U.throwIfError(await client.close());
+    if (metadata.error) fail();
+    return M.fromPrefixedHex(metadata.result);
+  } else if (path.isAbsolute(src)) {
+    return await loadMetadata(src);
+  } else {
+    try {
+      return await loadMetadata(path.fromFileUrl(src));
+    } catch (_e) {
+      unimplemented();
+    }
+  }
 }
 
-if (!args.metadata) throw new Error("Must specify metadata file");
-if (!args.output) throw new Error("Must specify output dir");
+async function loadMetadata(src: string) {
+  const ext = path.extname(src);
+  switch (ext) {
+    case ".scale": {
+      return M.fromPrefixedHex(await Deno.readTextFile(src));
+    }
+    case ".json": {
+      return unimplemented();
+    }
```

Comment:
The decoded metadata is both circular and has `Uint8Array`s etc., so I don't think we would ever support this.

Reply:
A json ext would indicate a chain spec (which I suppose we'll need to validate)